src/utils/get-server-status.ts

```diff
@@ -0,0 +1,181 @@
+
+        // If count is larger or equal than threshold...
+        // Set status to down
+        if (probeStatus.consecutiveFailures === threshold) {
+          log(`Send error for ${errorName}`)
+          probeStatus.shouldSendNotification = true
+          probeStatus.isDown = true
+        }
+      } else {
+        // Server is doing fine
+        probeStatus.shouldSendNotification = false
+        probeStatus.consecutiveFailures = 0
+        probeStatus.consecutiveSuccesses += 1
+        probeStatus.totalSuccesses += 1
+      }
+
+      return probeStatus
+    case true:
+      // If current probe is DOWN and trigger equals true
+      if (validation.status) {
+        // Server is not doing fine
+        probeStatus.shouldSendNotification = false
+        probeStatus.consecutiveSuccesses = 0
+        probeStatus.consecutiveFailures += 1
+        probeStatus.totalFailures += 1
+      } else {
+        // Add success count
+        probeStatus.shouldSendNotification = false
+        probeStatus.consecutiveFailures = 0
+        probeStatus.totalSuccesses += 1
+        probeStatus.consecutiveSuccesses += 1
+
+        // If count is larger or equal than threshold...
+        // Set status to UP
+        if (probeStatus.consecutiveSuccesses === threshold) {
+          log(`Send resolved for ${errorName}`)
+          probeStatus.shouldSendNotification = true
+          probeStatus.isDown = false
+        }
+      }
+
+      return probeStatus
+    default:
+      return probeStatus
+  }
+}
+
+export const getServerStatus = ({
+  probe,
+  validatedResp,
+  threshold,
+}: {
+  config: Config
+  probe: Probe
+  validatedResp: ValidateResponseStatus[]
+  threshold: number
+}) => {
+  try {
+    // Get Probe ID and Name
+    const { id, name } = probe
+    const results: Array<Status> = []
+
+    // Initialize server status
+    // This checks if there are no item in SERVER_STATUSES
+    // that doesn't have item with ID === id, push new ServerStatus
+    if (!SERVER_STATUSES.find((item) => item.id === id)) {
+      SERVER_STATUSES.push({
+        id,
+        name,
+        requests: 0,
+        responseTimeError: {
+          isDown: false,
+          shouldSendNotification: false,
+          totalSuccesses: 0,
+          totalFailures: 0,
+          consecutiveSuccesses: 0,
+          consecutiveFailures: 0,
+        },
+        statusCodeError: {
+          isDown: false,
+          shouldSendNotification: false,
+          totalSuccesses: 0,
+          totalFailures: 0,
+          consecutiveSuccesses: 0,
+          consecutiveFailures: 0,
+        },
+      })
+    }
+
+    // Check if there is any alert that is triggered
+    // If the alert is being triggered <threshold> times, send alert and
+    // change the server status respectively.
+    const probeIndex = SERVER_STATUSES.findIndex((probe) => probe.id === id)
+    const currentProbe = SERVER_STATUSES[probeIndex]
+
+    // Add request count
+    currentProbe.requests += 1
```

Comment:
This is one of the side effect, mutating data by adding counter.
Can we rename the function instead?

Reply:
Sure